#20 postgresql connect and queries execution

kortov opened the pull request:
#20 : added functionality for database connection and for simple queries execution

kortov committed:
#20: added basic database connection functionality

- added maven dependencies
- added code for connection to a database

kortov committed:
#20: implemented all required functionality

- printing list of tables
- creating user table
- inserting into user table
- printing user table content
- changing user password
- removing user from table

kortov committed:
#20: added missing jdbc class loader and minor improvements

- removed redundant newlines

kortov committed:
#20: added tests

kortov committed:
#20: improved printTable functionality: order by id

- repaired test correspondingly

ivshapovalov commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,7 +1,50 @@
+                + "JDBC Connection Testing ------------");
+
+        try {
+
```
Too many blank lines, I think.

kortov replied:
Were fixed in previous commits

ivshapovalov commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,7 +1,50 @@
+
+        try {
+
+            connection = DriverManager.getConnection(
```
Connection have to be closed

ivshapovalov commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,50 +1,223 @@
 public class Application {
+
+    private static final String JDBC_SQL_DRIVER = "org.postgresql.Driver";
+    private static final String JDBC_PROTOCOLS = "jdbc:postgresql://";
```
Maybe singular noun?

kortov replied:
I am not so sure.. jdbc is a protocol, postgresql is a subprotocol as far as I know

kortov replied:
Fixed it anyway

ivshapovalov commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,50 +1,223 @@
+
+    private static final String JDBC_SQL_DRIVER = "org.postgresql.Driver";
+    private static final String JDBC_PROTOCOLS = "jdbc:postgresql://";
+    private static final String SQl_URL = "127.0.0.1:5432/";
```
Maybe combine  JDBC_PROTOCOLS and SQl_URL into single variable?

kortov replied:
Url and port may be changed without changing database, I think?

ivshapovalov commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,50 +1,223 @@
+    private static final String LINE_SEPARATOR = System.getProperty("line.separator");
+
+    private Connection connection = null;
+
```
For what this blank lines?

kortov replied:
Just for readability, I'll remove them in the next PR

ivshapovalov commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,50 +1,223 @@
+    }
 
-        System.out.println("PostgreSQL JDBC Driver Registered!");
+    private void printStatusMsg(String operation, String subject) {
```
printStatus is enough for method name

kortov replied:
Will be fixed next PR

ivshapovalov commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,50 +1,223 @@
+        printMsgAffectedRows(affectedRows);
+    }
+
+    private void removeUserFromTable(String tableName, String value) {
```
Method name include 'User' word, but tableName in parameters.

kortov replied:
It needs to be corrected but no sure how exactly)

kortov replied:
Will be fixed next PR

ivshapovalov commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,50 +1,223 @@
-                    "sqlcmd");
+    private void printMsgAffectedRows(int affectedRows) {
+        if (affectedRows < 0) {
+            System.out.println("The query ended unsuccessfully, 0 rows affected");
```
Why unsuccessfully? If table contain 0 rows and query return 0 rows - I think that is successfull and correct result.

kortov replied:
Pay attention to “< 0”, please. But I agree that it could be difficult logic ( not an OOP way).

ivshapovalov replied:
Ou sure.

ivshapovalov commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,50 +1,223 @@
+        }
+    }
+
+    private void truncate(StringBuilder stringBuilder, int charsCount) {
```
I thought originally this is method for Truncate table:) Why you use separate method instead of use line
stringBuilder.delete(stringBuilder.length() - charsCount, stringBuilder.length()); 
in cycle?

kortov replied:
Ok, I’ll check it tonight

kortov replied:
Originally it was intended for readability: it is better to read method's name than awkward line of code. I don't like writing comments neither. But I didn't see that method contained only 1 line of code)

ivshapovalov commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -10,7 +10,7 @@
 public class Application {
 
-    private static final String JDBC_SQL_DRIVER = "org.postgresql.Driver";
+    private static final String JDBC_POSTGRESSQL_DRIVER = "org.postgresql.Driver";
```
One S in 'postresql' word

kortov replied:
Will be fixed next PR

ivshapovalov commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -99,7 +99,7 @@ private String tableNameFormatted(String tableName) {
 
     private void printTable(String tableName) {
-        String sqlQuery = "SELECT * FROM " + tableNameFormatted(tableName);
+        String sqlQuery = String.format("SELECT * FROM %1$s order by %1$s.id", tableNameFormatted(tableName));
```
We have one parameter in String.format(). Why you use this pattern?

kortov replied:
It uses same value twice, I remember, it didn’t compile yesterday with one parameter. It need to be checked

kortov replied:
It uses same value twice, I remember, it didn’t compile yesterday with one parameter. It need to be checked

kortov replied:
There are two ways to use one parameter:
1) String.format("%s %<s", "hello")
2) String.format("%1$s %1$s", hello);

Shtramak commented on pom.xml:
```diff
@@ -16,12 +16,19 @@
     </properties>
 
     <dependencies>
+        <!-- https://mvnrepository.com/artifact/junit/junit -->
```
Why you need this line?

kortov replied:
Just for clarity. Okay, removed it.

Shtramak commented on pom.xml:
```diff
@@ -16,12 +16,19 @@
     </properties>
 
     <dependencies>
+        <!-- https://mvnrepository.com/artifact/junit/junit -->
```
Why you need this line?

Shtramak replied:
I don't know why this comment appears 2 times... Probably gitHub decides that it's very important )))

kortov replied:
Just for clarity. Okay, removed it.

Shtramak commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,7 +1,218 @@
+
+    private static final String JDBC_POSTGRESSQL_DRIVER = "org.postgresql.Driver";
+    private static final String JDBC_PROTOCOLS = "jdbc:postgresql://";
+    private static final String SQl_URL = "127.0.0.1:5432/";
```
Written mistake. Should be SQL_URL

kortov replied:
Fixed

Shtramak commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,7 +1,218 @@
+    private void dropTableifExists(String tableName) {
+        String sqlQuery = "DROP TABLE IF EXISTS " + tableNameFormatted(tableName);
+        printStatusMsg("Removing", tableName);
+        int affectedRows = executeQuery(sqlQuery);
```
If you<s>'ll</s> have an Exception during the "int affectedRows = executeQuery(sqlQuery)" in out it will be StackTrace + "The query ended unsuccessfully, 0 rows affected". If you're providing a control under the query execution, then StackTrace is not good idea. My advice is to delete  "e.printStackTrace()" from catch block in executeQuery() method

kortov replied:
Followed your advice, but the codeflow needs to be reinvented

Shtramak commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,7 +1,218 @@
+    private void printMsgAffectedRows(int affectedRows) {
+        if (affectedRows < 0) {
+            System.out.println("The query ended unsuccessfully, 0 rows affected");
+            System.out.println();
```
You could use System.out.println("The query ended unsuccessfully, 0 rows affected"+LINE_SEPARATOR) instead of this line

kortov replied:
Fixed

Shtramak commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,7 +1,218 @@
+            ResultSet resultSet = statement.getResultSet();
+
+            if (!resultSet.isBeforeFirst()) {
+                System.out.println("table is empty");
```
Could be System.out.println("table is empty"+LINE_SEPARATOR)

kortov replied:
Fixed

Shtramak commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,7 +1,218 @@
+
+    private Connection connection = null;
+
+    static {
```
Why not to move this block to getConnection() method? If it will be no JDBC driver than you'll have simply 2 prints of stacktraces: first from this block and second from getConnection(). getConnection() is fully responsible for connection to DB and missing of the JDBC driver is one of the options

kortov replied:
Will be fixed next PR

Shtramak commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,7 +1,218 @@
+        try {
+            printStatusMsg("Printing", "list");
+            DatabaseMetaData md = connection.getMetaData();
+            ResultSet resultSet = md.getTables(null, "public", "%", new String[]{"TABLE"});
```
ResultSet is a resourse and needs to be closed (http://bit.ly/2zHWGEE)

kortov replied:
Will be fixed next PR

kortov committed:
#20: fixed most of the reviewers' comments

- fixed exceptions handling and codeflow
- added closing ResultSet resourses
- fixed typos in constants names
- removed redundant newlines from constants declarations
- fixed methods’ names
- removed redundant methods
- removed redundant lines from pom.xml

AVM82 commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,7 +1,190 @@
+            statement.execute();
+            ResultSet resultSet = statement.getResultSet();
+
+            if (!resultSet.isBeforeFirst()) {
```
Maybe better write like this:
if (resultSet.isBeforeFirst()) {
...
} else { 
System.out.println("table is empty" + LINE_SEPARATOR);
}
What do you think?

AVM82 commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,7 +1,190 @@
+    private void printTable(String tableName) {
+        String sqlQuery = String.format("SELECT * FROM %s order by %<s.id", USER_TABLE_NAME);
+        try (PreparedStatement statement = connection.prepareStatement(sqlQuery)) {
+            printStatus("Printing", tableName);
```
Why do you use PreparedStatement?
I use PreparedStatement when passing parameters, else Statement.
[https://www.javatpoint.com/PreparedStatement-interface](https://www.javatpoint.com/PreparedStatement-interface)

adaleorion replied:
I agree with @AVM82

AVM82 commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,7 +1,190 @@
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+        return -1;
```
Why "-1"?
Magic number...

AVM82 reviewed (approved):
The implementation is a little redundant, as for me, but overall it's look good at me

adaleorion commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,7 +1,190 @@
+        removeFromUserTable("user3");
+        printTable(USER_TABLE_NAME);
+
+        if (connection != null) {
```
I think you cannot close connection because you don't create it.

adaleorion commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,7 +1,190 @@
+    private Connection connection = null;
+
+    Application() throws SQLException, ClassNotFoundException {
+        connection = getConnection();
```
Connection should create in simpleSQL() 
if you close it there

adaleorion commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,7 +1,190 @@
+    }
+
+    private void changePasswordInUserTable(String name, String password) {
+        String sqlQuery = String.format("UPDATE " + USER_TABLE_NAME + " SET password = '%s'" + " WHERE name='%s'", password, name);
```
I think you should use prepared statement for this case

adaleorion commented on src/main/java/juja/sqlcmd/Application.java:
```diff
@@ -1,7 +1,190 @@
+        try (PreparedStatement statement = connection.prepareStatement(sqlQuery)) {
+            return statement.executeUpdate();
+        } catch (SQLException e) {
+            // gotcha
```
It's bad style. 
Why you catch exception and do nothing.
You should catch it in high level, for example in your main method

kortov committed:
#20: simplified logic and reorganized exceptions handling

- removed printStatus and printAffectedRows methods
- repaired test
- all sql exceptions are throwable to the next level of code flow

kortov committed:
#20:  improved exceptions handling and resources releasing

- removed try/catch from printing methods
- moved ResultSet initialisation into try-with-resources block
